packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,761 @@
+///                activeIcon: Icon(Icons.favorite),
+///                label: Text('First'),
+///              ),
+///              NavigationRailDestination(
+///                icon: Icon(Icons.bookmark_border),
+///                activeIcon: Icon(Icons.book),
+///                label: Text('Second'),
+///              ),
+///              NavigationRailDestination(
+///                icon: Icon(Icons.star_border),
+///                activeIcon: Icon(Icons.star),
+///                label: Text('Third'),
+///              ),
+///            ],
+///            onDestinationSelected: (int index) {
+///              setState(() {
+///                _currentIndex = index;
+///              });
+///            },
+///          ),
+///          VerticalDivider(thickness: 1, width: 1),
+///          Expanded(
+///            child: Center(
+///              child: Text('currentIndex: $_currentIndex'),
+///            ),
+///          )
+///        ],
+///      ),
+///    );
+///  }
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Scaffold], which can display the navigation rail within a [Row] of the
+///    [Scaffold.body] slot.
+///  * [NavigationRailDestination], which is used as a model to create tappable
+///    destinations in the navigation rail.
+///  * [BottomNavigationBar], which is used as a horizontal alternative for
+///    the same style of navigation as the navigation rail.
+class NavigationRail extends StatefulWidget {
+  /// Creates a material design navigation rail.
+  ///
+  /// The argument [destinations] must not be null. Additionally, it must be
+  /// non-empty.
+  ///
+  /// If [elevation] is specified, it must be non-negative.
+  ///
+  /// If [preferredWidth] is specified, it must be non-negative, and if
+  /// [extendedWidth is specified, it must be non-negative and greater than
+  /// [preferredWidth].
+  ///
+  /// The argument [extended] must not be null. [extended] can only be set to
+  /// true when when the [labelType] is null or [NavigationRailLabelType.none].
+  ///
+  /// If [backgroundColor], [elevation], [groupAlignment], [labelType],
+  /// [unselectedLabelTextStyle], [unselectedLabelTextStyle],
+  /// [unselectedIconTheme], or [selectedIconTheme] are null, then their
+  /// [NavigationRailThemeData] values will be used. If the corresponding
+  /// [NavigationRailThemeData] property is null, then the navigation rail
+  /// defaults are used.
+  ///
+  /// Typically used within a [Row] of the [Scaffold.body] property.
+  NavigationRail({
+    this.backgroundColor,
+    this.extended = false,
+    this.leading,
+    this.trailing,
+    @required this.destinations,
+    this.currentIndex = 0,
+    this.onDestinationSelected,
+    this.elevation,
+    this.groupAlignment,
+    this.labelType,
+    this.unselectedLabelTextStyle,
+    this.selectedLabelTextStyle,
+    this.unselectedIconTheme,
+    this.selectedIconTheme,
+    this.preferredWidth = _railWidth,
```

Comment:
The spec wants the rail to default to a width of 72, so that even when the content would not take up the whole width, it would still be 72. Since there is also a concept of a "compact" rail with a width of 56, I think it makes sense to be able to control this width. However, maybe minWidth is a better name. The rail still grows for longer labels, larger font sizes, or larger text scales.